embrace-android-sdk/src/main/java/io/embrace/android/embracesdk/comms/delivery/EmbracePendingApiCallsSender.kt

```diff
@@ -163,49 +130,104 @@
+                                    response.endpoint,
+                                    response.retryAfter,
+                                    this::scheduleApiCallsDelivery
+                                )
+                            }
+                            is ApiResponse.Incomplete -> {
+                                applyExponentialBackoff = true
+                            }
+                            else -> {
+                                // Not expected
+                            }
+                        }
+                        // Should retry, so we add it back to the queue.
+                        failedApiCallsToRetry.add(pendingApiCall)
+                    } else {
+                        // Shouldn't retry, so save the pending api calls without the removed one.
+                        cacheManager.savePendingApiCalls(pendingApiCalls)
+                    }
+                }
+            }
+
+            // Add back to the queue all retries that failed.
+            failedApiCallsToRetry.forEach {
+                pendingApiCalls.add(it)
+            }
+
+            if (pendingApiCalls.hasPendingApiCallsToSend()) {
+                scheduledExecutorService.submit {
+                    scheduleNextApiCallsDelivery(
+                        applyExponentialBackoff,
+                        delayInSeconds
+                    )
+                }
+            }
+        } catch (ex: Exception) {
+            logger.logDebug("Error when sending API call", ex)
+        }
+    }
 
     /**
      * Send the request for a PendingApiCall.
      */
-    private fun sendPendingApiCall(call: PendingApiCall): Boolean {
+    private fun sendPendingApiCall(call: PendingApiCall): ApiResponse? {
         val payload = cacheManager.loadPayload(call.cachedPayloadFilename)
-        if (payload != null) {
-            try {
-                logger.logDeveloper(TAG, "Sending a Pending API call")
-                sendMethod(call.apiRequest, payload)
-                cacheManager.deletePayload(call.cachedPayloadFilename)
-            } catch (ex: Exception) {
-                logger.logDeveloper(
-                    TAG,
-                    "Sending the Pending Api Call failed, scheduling to retry later",
-                    ex
-                )
-                return false
-            }
-        } else {
-            logger.logError("Could not retrieve cached api payload")
-            // If payload is null, the file could have been removed.
-            // We don't need to retry sending in the future as we'd get the same result.
-            // That's the reason for returning true.
+
+        if (payload == null) {
+            // If payload is null, the file could have been removed. We don't have to retry this call.
+            logger.logDeveloper(TAG, "Could not retrieve cached api payload")
+            return null
         }
-        return true
+        return sendMethod(call.apiRequest, payload)
```

Comment:
Do we need to check the response and delete it from the cache if we're not going to retry?

Reply:
Absolutely... I lost the deletion of the payload on the refactor. Updated.